pkg/controller/deployment/util/deployment_util.go

```diff
@@ -60,12 +60,119 @@ const (
+	// TimedOutReason is added in a deployment when its newest replica set fails to show any progress
+	// within the given deadline (progressDeadlineSeconds).
+	TimedOutReason = "DeploymentTimedOut"
+	// PausedDeployReason is added in a deployment when it is paused. Lack of progress shouldn't be
+	// estimated once a deployment is paused.
+	PausedDeployReason = "DeploymentPaused"
+	// ResumedDeployReason is added in a deployment when it is resumed. Useful for not failing accidentally
+	// deployments that paused amidst a rollout and are bounded by a deadline.
+	ResumedDeployReason = "DeploymentResumed"
 )
 
+// NewDeploymentCondition creates a new deployment condition.
+func NewDeploymentCondition(condType extensions.DeploymentConditionType, status api.ConditionStatus, reason, message string) *extensions.DeploymentCondition {
+	return &extensions.DeploymentCondition{
+		Type:               condType,
+		Status:             status,
+		LastTransitionTime: unversioned.Now(),
+		Reason:             reason,
+		Message:            message,
+	}
+}
+
+// DeploymentConditionExists returns if a condition with the provided type, status, and reason already exists
+// in the deployment status.
+func DeploymentConditionExists(status extensions.DeploymentStatus, condType extensions.DeploymentConditionType, condStatus api.ConditionStatus, condReason string) bool {
+	for _, c := range status.Conditions {
+		if c.Type == condType && c.Status == condStatus && c.Reason == condReason {
+			return true
+		}
+	}
+	return false
+}
+
+// SetDeploymentCondition updates the deployment to include the provided condition.
+func SetDeploymentCondition(status *extensions.DeploymentStatus, condition extensions.DeploymentCondition) {
+	newConditions := filterOutCondition(status.Conditions, condition.Type)
+	status.Conditions = append(newConditions, condition)
+}
+
+// RemoveDeploymentCondition removes the deployment condition with the provided type.
+func RemoveDeploymentCondition(status *extensions.DeploymentStatus, condType extensions.DeploymentConditionType) {
+	status.Conditions = filterOutCondition(status.Conditions, condType)
+}
+
+// filterOutCondition returns a new slice of deployment conditions without conditions with the provided type.
+func filterOutCondition(conditions []extensions.DeploymentCondition, condType extensions.DeploymentConditionType) []extensions.DeploymentCondition {
+	newConditions := []extensions.DeploymentCondition{}
+	for _, c := range conditions {
+		if c.Type == condType {
+			continue
+		}
+		newConditions = append(newConditions, c)
+	}
+	return newConditions
+}
+
+// ConditionFromWarningEvent transforms a warning event into a deployment condition.
+// Events passed in this helper should belong to a replica set. Only warnings for a
+// replica set currently are for failed creation or deletion of a pod, so we bound
+// this helper to accept just those.
+func ConditionFromWarningEvent(event api.Event) *extensions.DeploymentCondition {
```

Comment:
Same TODO on the godoc here about how this is really a job of the RC to provide this.